Add ppc64le wheel build support

sandeepgupta12 opened the pull request:
This PR adds support for building ppc64le wheels as part of the CI/CD pipeline. The goal is to enable ppc64le architecture compatibility for wheel builds, ensuring that TensorFlow/PyArrow (or any related package) can be distributed for Power architecture users.

**Changes Introduced**
✅ Enabled ppc64le architecture in CI/CD pipeline 
✅ Added Dockerfile and build script to support ppc64le builds
✅ Modified .github\workflows\_linux-build.yml and added ppc64le.yml configuration to include ppc64le
✅ Added files to configure ephemeral self-hosted runner similar to s390x under .github\scripts\ppc64le-ci

**Motivation & Context**
Currently, ppc64le wheels are not built in the official release pipeline, requiring users to build them manually. By adding native support in CI, ensures that Power users can install the package seamlessly via pip

Additional Notes
- This is the first PR introducing ppc64le wheel builds; further optimizations can be made based on feedback.
- Open to suggestions on improving the build process, dependencies, or CI/CD efficiency.
 -We can assist in setting up and maintaining the ephemeral self-hosted runner on an OSU VM if needed.

**Creation of OSU VM**: To facilitate further testing and CI integration, we request the creation of an OSU VM configured for PPC64LE. Below are the details where you can create the OSU VM-
URL- https://osuosl.org/services/powerdev/request_hosting/
IBM Advocate- [gerrit@us.ibm.com](mailto:gerrit@us.ibm.com)

**Details:**
The Open Source Lab (OSL) at Oregon State University (OSU), in partnership with IBM, provides access to IBM Power processor-based servers for developing and testing open source projects. The OSL offers following clusters:
OpenStack (non-GPU) Cluster:
• Architecture: Power little endian (LE) instances
• Virtualization: Kernel-based virtual machine (KVM)
• Access: Via Secure Shell (SSH) and/or through OpenStack's API and GUI interface
• Capabilities: Ideal for functional development and continuous integration (CI) work. It supports a managed Jenkins service hosted on the cluster or as a node incorporated into an external CI/CD pipeline.

**Additional Information:**
• We are prepared to provide any further details or assistance needed to support the PPC64LE architecture.

Please let us know if there are any specific requirements or steps needed to move forward with this request.

sandeepgupta12 committed:
Add ppc64le wheel build support

sandeepgupta12 committed:
Optimize build performance and upgrade base image to UBI 9.5

sandeepgupta12 commented:
I have signed the EasyCLA, and the check has be passing now. Please let me know if any further action is needed. Looking forward to your review and feedback on this change.

huydhn commented on .github/workflows/ppc64le.yml:
```diff
@@ -0,0 +1,37 @@
+jobs:
+  linux-ppc64le-docker-image-build:
+    name: Build docker image for ppc64le
+    runs-on: linux.ppc64le 
```
Do you know where this runner comes from?  Have it been added to PyTorch yet?

sandeepgupta12 replied:
Hi @huydhn  ,

Thank you for your review! The linux.ppc64le runner refers to a self-hosted runner for the PPC64LE architecture. It has not been added to PyTorch yet, but we can assist in setting it up.

Currently, we are leveraging OSU’s PowerDev infrastructure, which provides Power-based VMs for open-source projects. We can help configure an ephemeral self-hosted runner similar to what’s available for s390x( https://github.com/pytorch/pytorch/blob/main/.github/scripts/s390x-ci/README.md).

The necessary files to set up the ephemeral self-hosted runner are included in this PR under this .github/scripts/ppc64le-ci/README.md.

Would you like us to coordinate setting up a PPC64LE runner within the PyTorch CI environment? Let me know how we can proceed.

Best,
Sandeep

AlekseiNikiforovIBM commented on .github/workflows/ppc64le.yml:
```diff
@@ -0,0 +1,37 @@
+          submodules: true
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - name: Build Docker image for ppc64le
+        run: |
+          docker build -f .ci/docker/manywheel/Dockerfile_ppc64le -t pytorch-ppc64le:ubi9.3 .
+
+  ppc64le-UBI-9-3-Build-and-Test:
+    name: ppc64le-UBI-9-3-Build-and-Test
+    uses: ./.github/workflows/_linux-build.yml
+    needs: linux-ppc64le-docker-image-build
+    with:
+      build-environment: linux-ppc64le-binary-manywheel-ubi9
+      docker-image-name: pytorch-ppc64le:ubi9.3
```
As far as I understand, similar to s390x setup, runners run in a container, and are reset between each job.

So, there is one job that builds container, but never publishes it. When runner finishes job and is reset, built docker image is removed. And next job requires that image being available somewhere. And it can even run on different runner if there are many runners.

So, if this image is publicly available somewhere already, then it's not necessary to build the image. If image is not available, then building image to just discard it won't help.

That's exactly why I'm building image in advance [1], and loading it as first action of runner [2]. This way I can have needed image even if it's not published.

It could be possible to build image and use it if everything is put into same job if you don't mind spend time building same image for each job.

[1] https://github.com/pytorch/pytorch/blob/303ad1916fbcf5ad477ff7a4b5c5abe46e1418a7/.github/scripts/s390x-ci/README.md?plain=1#L30-L52
[2] https://github.com/pytorch/pytorch/blob/303ad1916fbcf5ad477ff7a4b5c5abe46e1418a7/.github/scripts/s390x-ci/self-hosted-builder/fs/usr/bin/actions-runner#L5-L10

sandeepgupta12 replied:
Fix issue with Docker image load as per review

sandeepgupta12 committed:
removed permision to file

sandeepgupta12 commented on .ci/docker/manywheel/Dockerfile_ppc64le:
```diff
@@ -0,0 +1,44 @@
+RUN chmod +x /ppc64le-build.sh
+
+# Verify permissions and ensure Unix line endings
+RUN dos2unix /ppc64le-build.sh || sed -i 's/\r$//' /ppc64le-build.sh
```
Resolved: Removed unnecessary steps from Dockerfile

AlekseiNikiforovIBM commented on .github/workflows/_linux-build.yml:
```diff
@@ -119,16 +119,17 @@ jobs:
       # checkout. In other cases you should prefer a local checkout.
       - name: Checkout PyTorch
         uses: pytorch/pytorch/.github/actions/checkout-pytorch@main
+        if: inputs.build-environment != 'linux-ppc64le-binary-manywheel-ubi9'
```
Why are you skipping this step?

I don't see you checking out a commit specified by job. I only see you copying some state of pytorch repository here:

https://github.com/pytorch/pytorch/pull/147194/files#diff-9fc2d81d1bbe4d201db02bdd3acbc341e8cd9056960739dbedf8b8395e0054e1R29

I don't see you updating it or checking out relevant commit. It looks like you're always rebuilding same commit. Or did I miss something?

AlekseiNikiforovIBM replied:
Yes, you're right, repository is updated. Still, I think it should be updated in build step, not in building docker image step.

sandeepgupta12 replied:
Fix issue with Docker image load as per review. 

Update runner registration flow based on feedback

AlekseiNikiforovIBM commented on .github/workflows/_linux-build.yml:
```diff
@@ -153,24 +154,26 @@ jobs:
 
       - name: Parse ref
         id: parse-ref
+        if: inputs.build-environment != 'linux-ppc64le-binary-manywheel-ubi9'
```
Why is this step skipped?

sandeepgupta12 replied:
Skipping parse-ref is intentional

The linux-ppc64le-binary-manywheel-ubi9 build does not require parsing the ref, as the relevant data is already handled in earlier steps.

If needed, I will revert it.

AlekseiNikiforovIBM replied:
It is informational step, I think, so please do revert it.

sandeepgupta12 replied:
Reverted

AlekseiNikiforovIBM commented on .github/workflows/_linux-build.yml:
```diff
@@ -153,24 +154,26 @@ jobs:
         id: get-job-id
         uses: ./.github/actions/get-workflow-job-id
-        if: always()
+        if: always() && inputs.build-environment != 'linux-ppc64le-binary-manywheel-ubi9'
```
Same

AlekseiNikiforovIBM commented on .github/workflows/_linux-build.yml:
```diff
@@ -153,24 +154,26 @@ jobs:
       # Apply the filter logic to the build step too if the test-config label is already there
       - name: Select all requested test configurations (if the test matrix is available)
         id: filter
+        if: inputs.build-environment != 'linux-ppc64le-binary-manywheel-ubi9'
```
Same

AlekseiNikiforovIBM commented on .github/scripts/ppc64le-ci/self-hosted-builder/fs/usr/bin/actions-runner:
```diff
@@ -0,0 +1,75 @@
+token_file=registration-token.json
+
+# Function to clean up and unregister the runner
+cleanup() {
```
Is this cleanup function even needed?
Runner is ephemeral, so it is unregistered automatically after 1 job. And pytorch repository periodically also removes offline runners if runner for some reason registers but never runs.

sandeepgupta12 replied:
reverted

AlekseiNikiforovIBM commented on .github/workflows/_linux-build.yml:
```diff
@@ -181,14 +184,14 @@ jobs:
 
       - name: Build
-        if: steps.filter.outputs.is-test-matrix-empty == 'False' || inputs.test-matrix == ''
+        if: (steps.filter.outputs.is-test-matrix-empty == 'False' || inputs.test-matrix == '') && (inputs.build-environment != 'linux-ppc64le-binary-manywheel-ubi9')
```
Why is this build step not reused? This step is most important in this workflow. Without it there's not a lot of reason to even use this workflow file.

sandeepgupta12 replied:
reverted

AlekseiNikiforovIBM commented on .github/scripts/ppc64le-ci/self-hosted-builder/fs/usr/bin/actions-runner:
```diff
@@ -0,0 +1,75 @@
+fi
+
+
+ACCESS_TOKEN="$(cat /run/runner_secret)"
```
If /run/runner_secret is a regular file that is just filled with "cat" program, then it is possible to read this token again later, for example, from inside the job, and leak it.

By looking at this line [1] I can assume that it is a regular file. At least I didn't find anything that would contradict such assumption.

[1] https://github.com/pytorch/pytorch/pull/147194/files#diff-2407110dcda3ecfcfd093afe86089b15d31e77bfc9394c46fb2e3503fb293cabR15

sandeepgupta12 replied:
Security is maintained

- This approach is consistent with how self-hosted runners are set up for s390x.
- The token is only used to create the runner and is not exposed beyond that scope.
- Once the job completes, the runner and its container are destroyed, ensuring no persistent access to the token.
- To further prevent any potential token leakage, we can delete /run/runner_secret immediately after generating the token.This ensures the token is not accessible beyond its intended use, even within the job.Let me know if you’d like me to update the script accordingly!

AlekseiNikiforovIBM replied:
> This approach is consistent with how self-hosted runners are set up for s390x.

No. I use pipe for s390x, here is regular file used. That's the root of this issue.

> Once the job completes, the runner and its container are destroyed, ensuring no persistent access to the token.

But it is actually accessible while job runs. So it can be accessed from inside the job.

>     To further prevent any potential token leakage, we can delete /run/runner_secret immediately after generating the token.This ensures the token is not accessible beyond its intended use, even within the job.Let me know if you’d like me to update the script accordingly!

You cannot delete mounted file from the container. But you could overwrite it contents, for example.

sandeepgupta12 committed:
Address review comments:Fix issue with Docker image load as per review

Update runner registration flow based on feedback

Enhance security by unmounting and removing runner token file

To prevent any potential token leakage, unmount and remove /run/runner_secret immediately after generating the token. This ensures that the token is inaccessible beyond its intended use, even within the job execution.

sandeepgupta12 committed:
Address review comments: Permission changes and removed mount docker socket

sandeepgupta12 committed:
Reuse _linux-build.yml for ppc64le build and updated other files accordingly

sandeepgupta12 commented on .github/workflows/ppc64le.yml:
```diff
@@ -0,0 +1,49 @@
+  cancel-in-progress: true
+
+jobs:
+  ppc64le-UBI-9-Build-and-Test:
```
@AlekseiNikiforovIBM, I have used the common workfow file and updated other required files accordingly. Please review the changes.

sandeepgupta12 committed:
Merge branch 'main' into ppc64le-ci-wheel-build-v1

sandeepgupta12 committed:
Merge branch 'main' into ppc64le-ci-wheel-build-v1

sandeepgupta12 committed:
Merge branch 'pytorch:main' into ppc64le-ci-wheel-build-v1

sandeepgupta12 committed:
Merge branch 'pytorch:main' into ppc64le-ci-wheel-build-v1

sandeepgupta12 committed:
Enable nightly build upload of wheels to PyPI for ppc64le - Added logic to upload built wheels to PyPI during nightly builds. - Ensured compatibility with ppc64le architecture.

tsambari2 commented:
@sandeepgupta12 , @huydhn , @malfet , @seemethere, Please can you help review this PR?
